### DISCUSSION TITLE
Where is ceph `client.admin` auth created?

### DISCUSSION THREAD

**Discussion by microyahoo** [rook/rook#12057] [Q&A] [answered]
Where is ceph `client.admin` auth created? I only saw the steps to create `rook-ceph-mon` secret in the [rook code](https://github.com/rook/rook/blob/e72616e712d4ac80bc5772a47805c1429a47a975/pkg/operator/ceph/controller/cluster_info.go#L95).
```console
[root@node01 smd]# ceph auth get client.admin
exported keyring for client.admin
[client.admin]
        key = AQBMqSljJh8gDBAAmr4wT2LLFqh3PxxoZk7f+cHA==
        caps mds = "allow *"
        caps mgr = "allow *"
        caps mon = "allow *"
        caps osd = "allow *"
```

**Comment by travisn** [comment]
Is the `genSecret()` method what you're looking for?
https://github.com/rook/rook/blob/e72616e712d4ac80bc5772a47805c1429a47a975/pkg/operator/ceph/controller/cluster_info.go#L218

**Reply by microyahoo** [reply]
The `genSecret()` only invokes the `ceph-authtool` to generate keyring file, but not create ceph auth like below. I noticed the `osd.x` auth was created in [activateOSDOnNodeCode](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/cluster/osd/spec.go#L112), the `client.bootstrap-osd` auth was created in [createOSDBootstrapKeyring](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/daemon/ceph/osd/init.go#L41), the `mgr.id` was created in [generateKeyring](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/cluster/mgr/config.go#L81), `client.rgw` was created in [generateCephXUser](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/object/config.go#L95), `client.crash` was created in [createCrashCollectorKeyring](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/cluster/nodedaemon/keyring.go#L66), `client.rbd-mirror` was created in [generateKeyring](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/cluster/rbd/config.go#L52), `mds.id` was created in [generateKeyring](https://github.com/rook/rook/blob/85d7920b2ae9180594375da06d2c159d6087c5d2/pkg/operator/ceph/file/mds/config.go#L40), but I haven't found out where `client.admin` auth is created.
```console
🍺  /root/go/src/deeproute.ai/rook ☞ git:(master) ✗ ceph-authtool --create-keyring client.bar.keyring --gen-key -n client.bar --cap  mon 'allow *' --cap osd 'allow *' --cap mds 'allow *'
creating client.bar.keyring

🍺 /root/go/src/deeproute.ai/rook ☞ git:(master) ✗ cat client.bar.keyring
[client.bar]
        key = AQB3cjVkkqpGGxAAqVm+iw6pmtzD8skp2ZbfAw==
        caps mds = "allow *"
        caps mon = "allow *"
        caps osd = "allow *"

🍺 /root/go/src/deeproute.ai/rook ☞ git:(master) ✗ ceph auth get client.bar
Error ENOENT: failed to find client.bar in keyring
```

**Reply by travisn** [reply, marked as answer]
The admin and mon keys are created differently. When the mons first start, they are told what the admin and mon keyrings are for bootstrapping the cluster. There is no separate step for registering those two keyrings with ceph since they were required during bootstrapping of the mons.
